csrc/scheduler/resize.cpp

```diff
@@ -245,6 +245,78 @@ std::unique_ptr<HeuristicParams> ResizeScheduler::computeHeuristics(
   return params;
 }
 
+namespace {
+
+// Before propagating loop transformations of the ref tensor, all
+// tensors need their loop domains to be reachable from the ref by
+// just a backward traversal in the exact graph. This is to make
+// tranformations to be consistently propagated through resize-induced
+// cycles in the exact graph.
+//
+// There may be tensors that have further transformations that do not
+// exist in the reference tensor, e.g., a reshape of the
+// reference. Here, we transform those tensors so that they become
+// ready to get the reference transformations propagated to.
+void prepareForBackwardTransformPropagation(TensorView* ref_tv) {
+  Fusion* fusion = ref_tv->fusion();
+
+  // Tensors that appear before the ref should not need anything here
+  // as they should already be ready to get propagated from the ref
+  const auto all_dep_stmts = StmtSort::getStmtsTo({ref_tv});
+  const std::unordered_set<Statement*> all_dep_stmt_set{
+      all_dep_stmts.begin(), all_dep_stmts.end()};
+
+  if (std::ranges::all_of(fusion->allTvs(), [&](TensorView* tv) {
+        return all_dep_stmt_set.contains(tv);
+      })) {
+    return;
+  }
+
+  IdModel id_model(fusion);
+  const auto& graph = id_model.buildBroadcastGraph();
+
+  const auto ref_logical = graph.toGroups(ref_tv->getLogicalDomain());
+
+  std::vector<TensorView*> tvs_with_extra_transforms;
+
+  // Look for tensor whose logical domain is not rechable from ref by
+  // just a backward traversal
+  for (auto tv : fusion->allTvs()) {
+    if (all_dep_stmt_set.contains(tv)) {
+      continue;
+    }
+
+    const auto tv_logical =
+        graph.toGroups(TensorDomain::noBroadcasts(tv->getLogicalDomain()));
+
+    auto all_visited = ValGraphBFS::getExprGroupsBetween(
+                           graph,
+                           ref_logical,
+                           tv_logical,
+                           /*require_all_to_visited=*/false,
+                           Direction::Backward)
+                           .second;
+
+    if (all_visited) {
+      continue;
+    }
+
+    // This tensor is unreachable by just traversing backward
+    tvs_with_extra_transforms.push_back(tv);
+  }
+
+  if (tvs_with_extra_transforms.empty()) {
+    return;
+  }
+
+  // Make their loop domains look like the ref so that loop
+  // transformations of the ref can be propagated
+  scheduler_tools::scheduleLoopDomainsLike(
+      tvs_with_extra_transforms, ref_tv->getLogicalDomain());
```

Comment:
Out of curiosity. How do we know it's a safe to overwrite the loop domain here?

i.e. if the overwrite loop domain from ref_tv is *smaller* than the actual tv, we are effectively applying a false predicate here, which might lead to silent wrong results?

I'm assuming the resize scheduler has a check somewhere that ensures the coverage from ref to outputs?

Reply:
That's a good question. Since this is only for pointwise ops with no reduction, as long as the loop domain of the reference tensor is kept intact, all tensors should be also covered. If not, that seems to indicate the reference is not really a right reference. We found a couple of issues with finding right references, but as far as I can see, a reference tensor chosen by the current logic should guarantee all other tensors can be transformed to the same shape as the reference without losing any dimension.

That won't be the case with reductions, so that's one big difference when we extend this scheduling scheme to handle reductions.

Reply:
nitpick: I wonder if we should add a check looking for reduction op and the reasoning as a comment in this function?
Not pushing strongly for that though.

Reply:
Thanks for the suggestion. Added a comment (not here, but the first place where potentially shrinking propagations happen).